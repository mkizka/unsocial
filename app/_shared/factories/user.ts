import { defineUserFactory } from "@/_generated";

export const LocalUserFactory = defineUserFactory({
  defaultData: {
    host: "myhost.example.com",
  },
});

export const RemoteUserFactory = defineUserFactory({
  defaultData: async ({ seq }) => ({
    host: "remote.example.com",
    actorUrl: `https://remote.example.com/users/${seq}`,
    inboxUrl: `https://remote.example.com/users/${seq}/inbox`,
  }),
});
