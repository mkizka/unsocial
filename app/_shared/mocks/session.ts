import { userSessionService } from "@/_shared/user/services/userSessionService";

jest.mock("@/_shared/user/services/userSessionService");
export const mockedGetSessionUserId = jest.mocked(
  userSessionService.getSessionUserId,
);
// export const mockedGetSessionUser = jest.mocked(getSessionUser);
