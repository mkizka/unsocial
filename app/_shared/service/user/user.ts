import { cache } from "react";

import {
  findOrFetchUserById,
  findOrFetchUserByWebFinger,
} from "./findOrFetchUser";
import { findOrFetchUserByActor as _findOrFetchUserByActor } from "./findOrFetchUser";
import { parseUserKey } from "./parseUserKey";

export const findOrFetchUserByActor = cache(_findOrFetchUserByActor);

export const findOrFetchUserByKey = cache(async (userKey: string) => {
  const parsed = parseUserKey(userKey);
  if ("id" in parsed) {
    return findOrFetchUserById(parsed.id);
  }
  if ("preferredUsername" in parsed) {
    return findOrFetchUserByWebFinger(parsed);
  }
  return parsed;
});
