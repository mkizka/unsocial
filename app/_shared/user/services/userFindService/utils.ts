import type { User } from "@prisma/client";

import { env } from "@/_shared/utils/env";

export const shouldRefetch = (user: User) => {
  if (user.host === env.UNSOCIAL_HOST) {
    return false;
  }
  if (!user.lastFetchedAt) {
    return true;
  }
  const diff = Date.now() - user.lastFetchedAt.getTime();
  return diff >= 1000 * 60 * 60 * 3;
};
