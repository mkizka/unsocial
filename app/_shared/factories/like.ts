import { defineLikeFactory } from "@/_generated";

import { LocalNoteFactory } from "./note";
import { LocalUserFactory } from "./user";

export const LikeFactory = defineLikeFactory({
  defaultData: {
    user: LocalUserFactory,
    note: LocalNoteFactory,
  },
});
