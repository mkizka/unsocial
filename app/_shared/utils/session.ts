import type { User } from "@prisma/client";
import { redirect } from "next/navigation";
import type { NextAuthOptions } from "next-auth";
import NextAuth, { getServerSession } from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import { cache } from "react";

import { userService } from "@/_shared/service/user";

import { createLogger } from "./logger";
import { prisma } from "./prisma";

const logger = createLogger("next-auth");

const authOptions: NextAuthOptions = {
  callbacks: {
    async session({ session, token }) {
      session.user = {
        id: token.id,
      };
      return session;
    },
    async jwt({ token, user }) {
      if (user) {
        token.id = user.id;
      }
      return token;
    },
  },
  logger: {
    error(code, message) {
      logger.error(`${code} ${JSON.stringify(message)}`);
    },
    warn(code) {
      logger.warn(code);
    },
  },
  providers: [
    CredentialsProvider({
      // authorizeの型エラーを解消するために必要
      // フォーム生成はしないので値は空にしておく
      credentials: {},
      authorize: userService.authorize,
    }),
  ],
};

export const handler = NextAuth(authOptions);

const getSessionUserIdOrNull = cache(async () => {
  const session = await getServerSession(authOptions);
  return session?.user?.id ?? null;
});

const getSessionUserOrNull = cache(async () => {
  const sessionUserId = await getSessionUserIdOrNull();
  if (!sessionUserId) {
    return null;
  }
  const sessionUser = await prisma.user.findUnique({
    where: {
      id: sessionUserId,
    },
  });
  return sessionUser ?? null;
});

export const getSessionUserId = async <T extends boolean>(params?: {
  redirect: T;
}) => {
  const userId = await getSessionUserIdOrNull();
  if (!userId && params?.redirect) {
    redirect("/auth");
  }
  return userId as T extends true ? string : string | null;
};

export const getSessionUser = async <T extends boolean>(params?: {
  redirect: T;
}) => {
  const user = await getSessionUserOrNull();
  if (!user && params?.redirect) {
    redirect("/auth");
  }
  return user as T extends true ? User : User | null;
};
