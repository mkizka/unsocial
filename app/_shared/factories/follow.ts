import { defineFollowFactory } from "@/_generated";
import { prisma } from "@/_shared/utils/prisma";

import { LocalUserFactory, RemoteUserFactory } from "./user";

const factory = defineFollowFactory({
  defaultData: {
    follower: RemoteUserFactory,
    followee: LocalUserFactory,
  },
});

export const FollowFactory = {
  create: async (...args: Parameters<typeof factory.build>) => {
    const data = await factory.build(...args);
    return prisma.follow.create({
      data,
      include: {
        follower: true,
        followee: true,
      },
    });
  },
};
