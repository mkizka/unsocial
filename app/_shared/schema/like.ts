import { z } from "zod";

import { env } from "@/_shared/utils/env";

export const inboxLikeSchema = z.object({
  type: z.literal("Like"),
  actor: z.string().url(),
  object: z
    .string()
    .url()
    .refine((val) => new URL(val).host === env.UNSOCIAL_HOST, {
      message: "自ホストのノートではありません",
    }),
  content: z.string().optional(),
});

type LikeActivity = z.infer<typeof inboxLikeSchema>;
