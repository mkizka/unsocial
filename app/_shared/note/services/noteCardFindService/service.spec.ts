import { LikeFactory } from "@/_shared/factories/like";
import {
  LocalNoteFactory,
  NoteWithQUoteFactory,
  RemoteNoteFactory,
} from "@/_shared/factories/note";
import { LocalUserFactory } from "@/_shared/factories/user";
import { mockedGetSessionUserId } from "@/_shared/mocks/session";
import { prisma } from "@/_shared/utils/prisma";

import { noteCardFindService } from ".";

describe("noteCardFindService", () => {
  describe("findUniqueNoteCard", () => {
    test("カード表示のための形式にしたノートを返す", async () => {
      // arrange
      const data = await LocalNoteFactory.build();
      const note = await prisma.note.create({ data, include: { user: true } });
      // act
      const noteCard = await noteCardFindService.findUniqueNoteCard(note.id);
      // assert
      expect(noteCard).toEqualPrisma(
        expect.objectContaining({
          quotedBy: null,
          attachmentUrls: [],
          isMine: false,
          isLiked: false,
          isReposted: false,
          url: `/notes/${note.id}`,
          user: expect.objectContaining({
            displayUsername: `@${note.user.preferredUsername}`,
            url: `/@${note.user.preferredUsername}`,
          }),
        }),
      );
    });
    test("自分の投稿である場合はisMineをtrueにする", async () => {
      // arrange
      const userId = await LocalUserFactory.createForConnect();
      const note = await LocalNoteFactory.create({
        user: {
          connect: userId,
        },
      });
      mockedGetSessionUserId.mockResolvedValueOnce(note.userId);
      // act
      const noteCard = await noteCardFindService.findUniqueNoteCard(note.id);
      // assert
      expect(noteCard?.isMine).toBe(true);
    });
    test("いいね済みの場合はisLikeをtrueにする", async () => {
      // arrange
      const user = await LocalUserFactory.createForConnect();
      const like = await LikeFactory.create({
        user: {
          connect: user,
        },
      });
      mockedGetSessionUserId.mockResolvedValueOnce(like.userId);
      // act
      const noteCard = await noteCardFindService.findUniqueNoteCard(
        like.noteId,
      );
      // assert
      expect(noteCard?.isLiked).toBe(true);
    });
    test("リポスト済み", async () => {
      // arrange
      const noteWithQuote = await NoteWithQUoteFactory.create();
      mockedGetSessionUserId.mockResolvedValueOnce(noteWithQuote.userId);
      // act
      const noteCard = await noteCardFindService.findUniqueNoteCard(
        noteWithQuote.id,
      );
      // assert
      expect(noteCard?.isReposted).toBe(true);
    });
    test("他ホストのユーザーはURLを@user@example.comの形式にする", async () => {
      // arrange
      const note = await RemoteNoteFactory.create();
      mockedGetSessionUserId.mockResolvedValueOnce(note.userId);
      // act
      const noteCard = await noteCardFindService.findUniqueNoteCard(note.id);
      // assert
      expect(noteCard!.user.url).toBe(
        `/@${noteCard!.user.preferredUsername}@${noteCard!.user.host}`,
      );
    });
    test("quoteがある場合はquotedByを追加", async () => {
      // arrange
      const quote = await LocalNoteFactory.createForConnect();
      const data = await LocalNoteFactory.build({
        quote: {
          connect: quote,
        },
      });
      const quotedNote = await prisma.note.create({
        data,
        include: { user: true },
      });
      // act
      const noteCard = await noteCardFindService.findUniqueNoteCard(
        quotedNote.id,
      );
      // assert
      expect(noteCard?.quotedBy).toEqual({
        host: quotedNote.user.host,
        preferredUsername: quotedNote.user.preferredUsername,
        name: null,
        url: `/@${quotedNote.user.preferredUsername}`,
      });
    });
    test("ノートが無かった場合はgetSessionUserIdOrNullを呼ばない", async () => {
      // act
      const noteCard = await noteCardFindService.findUniqueNoteCard("noteId");
      // assert
      expect(mockedGetSessionUserId).not.toHaveBeenCalled();
      expect(noteCard).toEqual(null);
    });
  });

  describe("findManyNoteCards", () => {
    test("カード表示のための形式にしたノートを複数返す", async () => {
      // arrange
      await LocalNoteFactory.createList(11);
      // act
      const noteCards = await noteCardFindService.findManyNoteCards({
        count: 10,
      });
      // assert
      expect(noteCards).toHaveLength(10);
      for (const noteCard of noteCards) {
        expect(noteCard).toEqualPrisma(
          expect.objectContaining({
            quotedBy: null,
            attachmentUrls: [],
            isMine: false,
            isLiked: false,
            isReposted: false,
            url: `/notes/${noteCard!.id}`,
            user: expect.objectContaining({
              displayUsername: `@${noteCard!.user.preferredUsername}`,
              url: `/@${noteCard!.user.preferredUsername}`,
            }),
          }),
        );
      }
    });
    test("publishedAtでソートする", async () => {
      // arrange
      const notes = await LocalNoteFactory.createList([
        { publishedAt: new Date("2024-01-01T00:00:00Z") },
        { publishedAt: new Date("2024-01-03T00:00:00Z") },
        { publishedAt: new Date("2024-01-02T00:00:00Z") },
      ]);
      // act
      const noteCards = await noteCardFindService.findManyNoteCards({
        count: 10,
      });
      // assert
      expect(noteCards[0]!.id).toBe(notes[1]!.id);
      expect(noteCards[1]!.id).toBe(notes[2]!.id);
      expect(noteCards[2]!.id).toBe(notes[0]!.id);
    });
    test("sinceからuntilまでのノートを取得する", async () => {
      // arrange
      const notes = await LocalNoteFactory.createList([
        { publishedAt: new Date("2024-01-01T00:00:00Z") },
        { publishedAt: new Date("2024-01-02T00:00:00Z") },
        { publishedAt: new Date("2024-01-03T00:00:00Z") },
      ]);
      // act
      const noteCards = await noteCardFindService.findManyNoteCards({
        count: 10,
        since: new Date("2024-01-01T00:00:01Z"), // 一つ目の投稿の1秒後
        until: new Date("2024-01-02T12:59:59Z"), // 三つ目の投稿の1秒前
      });
      // assert
      expect(noteCards).toHaveLength(1);
      expect(noteCards[0]!.id).toBe(notes[1]!.id);
    });
    test("ノートが無かった場合はgetSessionUserIdOrNullを呼ばない", async () => {
      // act
      const noteCards = await noteCardFindService.findManyNoteCards({
        count: 10,
      });
      // assert
      expect(mockedGetSessionUserId).not.toHaveBeenCalled();
      expect(noteCards).toEqual([]);
    });
  });
});
