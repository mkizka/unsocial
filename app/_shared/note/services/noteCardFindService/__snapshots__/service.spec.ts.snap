// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`noteCardFindService findManyNoteCards 正常系 1`] = `
{
  "attachmentUrls": [],
  "attachments": [],
  "id": "noteId",
  "isLiked": true,
  "isMine": true,
  "likes": [
    {
      "id": "likeId",
      "userId": "userId",
    },
    {
      "id": "likeId2",
      "userId": "userId2",
    },
  ],
  "quotedBy": null,
  "replies": [],
  "replyTo": null,
  "url": "/notes/noteId",
  "user": {
    "displayUsername": "@preferredUsername@remote.example.com",
    "host": "remote.example.com",
    "preferredUsername": "preferredUsername",
    "url": "/@preferredUsername@remote.example.com",
  },
  "userId": "userId",
}
`;

exports[`noteCardFindService findUniqueNoteCard 自分の投稿かついいね済み 1`] = `
{
  "attachmentUrls": [],
  "attachments": [],
  "id": "noteId",
  "isLiked": true,
  "isMine": true,
  "likes": [
    {
      "id": "likeId",
      "userId": "userId",
    },
    {
      "id": "likeId2",
      "userId": "userId2",
    },
  ],
  "quotedBy": null,
  "replies": [],
  "replyTo": null,
  "url": "/notes/noteId",
  "user": {
    "displayUsername": "@preferredUsername@remote.example.com",
    "host": "remote.example.com",
    "preferredUsername": "preferredUsername",
    "url": "/@preferredUsername@remote.example.com",
  },
  "userId": "userId",
}
`;
